pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/GenericTransformFunction.java

```diff
@@ -0,0 +1,265 @@
+                "Unsupported data type " + paramType.getTypeName() + "for transform function " + getName());
+        }
+      }
+    }
+  }
+
+  @Override
+  public TransformResultMetadata getResultMetadata() {
+    Class returnType = _functionInvoker.getReturnType();
+    switch(returnType.getTypeName()) {
+      case "java.lang.Integer": return INT_SV_NO_DICTIONARY_METADATA;
+      case "java.lang.Long": return LONG_SV_NO_DICTIONARY_METADATA;
+      case "java.lang.Float": return DOUBLE_SV_NO_DICTIONARY_METADATA;
+      case "java.lang.Double": return DOUBLE_SV_NO_DICTIONARY_METADATA;
+      case "java.lang.String": return STRING_SV_NO_DICTIONARY_METADATA;
+      default:
+        throw new RuntimeException("Unsupported data type " + returnType.getTypeName() + "for transform function " + getName());
+    }
+  }
+
+  @SuppressWarnings("Duplicates")
+  @Override
+  public int[] transformToIntValuesSV(ProjectionBlock projectionBlock) {
+    if (_integerResult == null) {
+      _integerResult = new int[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+    transformValues(projectionBlock);
+    return _integerResult;
+  }
+
+  @SuppressWarnings("Duplicates")
+  @Override
+  public long[] transformToLongValuesSV(ProjectionBlock projectionBlock) {
+    if (_longResult == null) {
+      _longResult = new long[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+    transformValues(projectionBlock);
+    return _longResult;
+  }
+
+  @SuppressWarnings("Duplicates")
+  @Override
+  public float[] transformToFloatValuesSV(ProjectionBlock projectionBlock) {
+    if (_floatResult == null) {
+      _floatResult = new float[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+    transformValues(projectionBlock);
+    return _floatResult;
+  }
+
+  @SuppressWarnings("Duplicates")
+  @Override
+  public double[] transformToDoubleValuesSV(ProjectionBlock projectionBlock) {
+    if (_doubleResult == null) {
+      _doubleResult = new double[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+    transformValues(projectionBlock);
+    return _doubleResult;
+  }
+
+  @SuppressWarnings("Duplicates")
+  @Override
+  public String[] transformToStringValuesSV(ProjectionBlock projectionBlock) {
+    if (_stringResult == null) {
+      _stringResult = new String[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+    transformValues(projectionBlock);
+    return _stringResult;
+  }
+
+  private void transformValues(ProjectionBlock projectionBlock) {
+    int length = projectionBlock.getNumDocs();
+    int numNonLiteralArgs = _nonLiteralArgIndices.size();
+    Object[][] nonLiteralBlockValues = new Object[numNonLiteralArgs][];
+
+    transformNonLiteralArgsToValues(projectionBlock, numNonLiteralArgs, nonLiteralBlockValues);
+
+    //now invoke the actual function
+    for (int i = 0; i < length; i++) {
+      for (int k = 0; k < numNonLiteralArgs; k++) {
+        _args[_nonLiteralArgIndices.get(k)] = nonLiteralBlockValues[k][i];
+      }
+
+      Class returnType = _functionInvoker.getReturnType();
```

Comment:
it's better to duplicate this for every type event though this code looks better for readability.

the switch case on every invocation will hurt perf